File: frontend/src/utils/napPayout.test.ts
import { describe, expect, it } from 'vitest';
import { napPayout } from './napPayout';

describe('napPayout', () => {
  it('stakes the trick count either way for the numbered contracts', () => {
    expect(napPayout(2)).toEqual({ make: 2, fail: 2 });
    expect(napPayout(3)).toEqual({ make: 3, fail: 3 });
    expect(napPayout(4)).toEqual({ make: 4, fail: 4 });
  });

  // **ここだけ非対称。**達成 10 に対し、失敗しても相手が得るのは各 5。
  it('pays Nap ten to make and five apiece to beat', () => {
    expect(napPayout(5)).toEqual({ make: 10, fail: 5 });
  });

  it('stakes nothing on a pass', () => {
    expect(napPayout(0)).toBeNull();
  });
});

File: internal/infrastructure/games/nap_payout_test.go
package games_test

import (
	"os"
	"path/filepath"
	"regexp"
	"strconv"
	"testing"

	"github.com/yuta-yoshinaga/go_trumpcards/internal/domain"
)

// TestNapWebPayoutMatchesTheDomain guards the web's copy of the contract stakes
// against the values the domain actually pays out.
//
// `frontend/src/utils/napPayout.ts` spells the Nap asymmetry (10 to make, 5
// apiece to beat) as literals so the page can label its bid buttons. Change
// makeValue/failValue in `internal/domain/Nap.go` and the buttons would keep
// promising the old stake with nothing to notice (#5651) -- the same drift the
// Baccarat payout panel had in #5497.
func TestNapWebPayoutMatchesTheDomain(t *testing.T) {
	root := filepath.Join("..", "..", "..")
	path := filepath.Join(root, "frontend", "src", "utils", "napPayout.ts")
	src, err := os.ReadFile(path)
	if err != nil {
		t.Fatalf("read %s: %v", path, err)
	}

	// The Nap branch is the only literal pair in the file.
	m := regexp.MustCompile(`return \{ make: (\d+), fail: (\d+) \};`).FindStringSubmatch(string(src))
	if m == nil {
		t.Fatal("napPayout.ts no longer states the Nap stake as a literal pair")
	}
	wantMake, wantFail := domain.NapBidPayout(domain.NapBidNap)
	if m[1] != strconv.Itoa(wantMake) || m[2] != strconv.Itoa(wantFail) {
		t.Errorf("napPayout.ts Nap = %s/%s, want %d/%d", m[1], m[2], wantMake, wantFail)
	}

	// The numbered contracts stake their own trick count both ways; if that ever
	// stops being true the web's `{ make: contract, fail: contract }` is wrong.
	for _, bid := range []domain.NapBid{domain.NapBidTwo, domain.NapBidThree, domain.NapBidFour} {
		gotMake, gotFail := domain.NapBidPayout(bid)
		if gotMake != int(bid) || gotFail != int(bid) {
			t.Errorf("NapBidPayout(%v) = %d/%d, but napPayout.ts assumes %d/%d",
				bid, gotMake, gotFail, int(bid), int(bid))
		}
	}
}
